File: AI/middleware/concurrent_limit.py
"""
第四层: 限制并发行动 (ConcurrentActionLimitMiddleware)

问题场景:
- 过多智能体同时执行行动,导致资源争抢
- 系统负载过高

解决方案:
- 限制同时执行的行动数量
"""

from typing import Any

from .base import BaseMiddleware, MiddlewareContext, MiddlewareResult


class ConcurrentActionLimitMiddleware(BaseMiddleware):
    """限制并发行动中间件"""

    def __init__(self, max_concurrent: int = 10):
        super().__init__("ConcurrentActionLimitMiddleware")
        self.max_concurrent = max_concurrent
        self._current_actions = 0

    def process(self, context: MiddlewareContext, agent: Any) -> MiddlewareResult:
        if not self.enabled:
            return MiddlewareResult()

        if context.phase.value == "before_action":
            if self._current_actions >= self.max_concurrent:
                return MiddlewareResult(
                    should_continue=False,
                    message=f"[{self.name}] 并发行动数已达上限 ({self.max_concurrent})"
                )
            self._current_actions += 1

        elif context.phase.value == "after_action":
            self._current_actions = max(0, self._current_actions - 1)

        return MiddlewareResult()

    def reset_step(self):
        """每个时间步重置并发计数"""
        self._current_actions = 0

File: src/AI/middleware/concurrent_limit.ts
/**
 * 并发限制中间件
 * 用于控制智能体的并发任务数，防止系统过载
 */
import { BaseMiddleware, MiddlewareContext, MiddlewareResult } from './base';

export class ConcurrentLimitMiddleware extends BaseMiddleware {
  // 最大并发数
  private maxConcurrent: number;
  // 当前并发数
  private currentConcurrent: number;

  /**
   * 构造函数
   * @param maxConcurrent 最大并发数，默认为3
   */
  constructor(maxConcurrent: number = 3) {
    super("ConcurrentLimitMiddleware");
    this.maxConcurrent = maxConcurrent;
    this.currentConcurrent = 0;
  }

  /**
   * 处理方法
   * 检查并限制并发任务数
   * @param context 中间件上下文
   * @param agent 智能体实例
   * @returns 中间件结果
   */
  process(context: MiddlewareContext, agent: any): MiddlewareResult {
    // 检查当前并发数是否达到上限
    if (this.currentConcurrent >= this.maxConcurrent) {
      return {
        should_continue: false,
        message: "并发任务数已达到上限，请稍后再试",
        metadata: {}
      };
    }

    // 增加当前并发数
    this.currentConcurrent++;
    return {
      should_continue: true,
      metadata: {}
    };
  }

  /**
   * 释放并发资源
   * 当任务完成时调用此方法减少当前并发数
   */
  release(): void {
    if (this.currentConcurrent > 0) {
      this.currentConcurrent--;
    }
  }
}
